torch/lib/TensorLib/templates/TensorDerived.cpp

```diff
@@ -0,0 +1,67 @@
+#include "TensorLib/${Tensor}.h"
+#include "TensorLib/HalfConvert.h"
+
+//sketch:
+
+/*
+
+class Type {
+  // ctor is configurable as well
+  Tensor new(...) {}
+  // all  methods and pure functions, virtually dispatched
+  virtual Tensor add(Tensor a, Tensor b);
+
+  // no Tensor arguments but virtually dispatched on type
+  virtual Tensor linspace(int a, int b);
+  //
+}
+
+// base class of derived tensors
+class TensorImpl {
+  Type * type_; //non-virtual so that we don't double indirect unnecessarily ...
+
+  // not virtual - Type * handles the virtual dispatch
+  Tensor add(Tensor b) {
+    type_->add(self,b);
+  }
+}
```

Comment:
Things like zeros/ones/linespace fit naturally as methods on Type because it combines both the backend and scalar type, which is enough to choose the concrete return type of the Tensor being created. It is also useful to pass them around to allow polymorphic construction of Tensors, making it easier to write cross-backend and cross-type code. Currently all dynamic dispatch goes through the Type objects.